File: bindings/javascript/tests/kv.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { AsyncDatabase, Database, type KvEntry } from "../js/index.js";

const temporaryDirectories: string[] = [];

function temporaryDatabasePath(): string {
  const directory = mkdtempSync(join(tmpdir(), "zova-javascript-kv-"));
  temporaryDirectories.push(directory);
  return join(directory, "test.zova");
}

function bytes(value: string | number[]): Uint8Array {
  if (typeof value === "string") {
    return new TextEncoder().encode(value);
  }
  return new Uint8Array(value);
}

afterEach(() => {
  for (const directory of temporaryDirectories.splice(0)) {
    rmSync(directory, { force: true, recursive: true });
  }
});

describe("Database key-value storage", () => {
  test("crud preserves exact bytes", () => {
    const database = Database.create(temporaryDatabasePath());
    try {
      database.kvPut(bytes("settings"), bytes("theme"), bytes("dark"));
      database.kvPut(bytes("settings"), bytes("retries"), bytes([0, 1, 2]));
      database.kvPut(bytes("settings"), bytes("empty"), new Uint8Array(0));

      expect(database.kvGet(bytes("settings"), bytes("theme"))).toEqual(bytes("dark"));
      expect(database.kvGet(bytes("settings"), bytes("retries"))).toEqual(
        bytes([0, 1, 2]),
      );
      expect(database.kvGet(bytes("settings"), bytes("empty"))).toEqual(
        new Uint8Array(0),
      );
      expect(database.kvGet(bytes("settings"), bytes("ghost"))).toBeNull();
      expect(database.kvGet(bytes("other"), bytes("theme"))).toBeNull();

      expect(database.kvCount(bytes("settings"))).toBe(3n);

      database.kvPut(bytes("settings"), bytes("theme"), bytes("light"));
      expect(database.kvGet(bytes("settings"), bytes("theme"))).toEqual(bytes("light"));
      expect(database.kvCount(bytes("settings"))).toBe(3n);

      database.kvDelete(bytes("settings"), bytes("theme"));
      expect(database.kvGet(bytes("settings"), bytes("theme"))).toBeNull();
      expect(database.kvCount(bytes("settings"))).toBe(2n);

      database.kvDelete(bytes("settings"), bytes("ghost"));
      expect(database.kvCount(bytes("settings"))).toBe(2n);

      database.kvClearNamespace(bytes("settings"));
      expect(database.kvCount(bytes("settings"))).toBe(0n);
    } finally {
      database.close();
    }
  });

  test("get many preserves order and duplicates", () => {
    const database = Database.create(temporaryDatabasePath());
    try {
      database.kvPut(bytes("ns"), bytes("a"), bytes("1"));
      database.kvPut(bytes("ns"), bytes("b"), bytes("2"));
      database.kvPut(bytes("ns"), bytes("c"), bytes("3"));

      const results = database.kvGetMany(bytes("ns"), [
        bytes("c"),
        bytes("ghost"),
        bytes("a"),
        bytes("c"),
      ]);
      expect(results.length).toBe(4);
      expect(results[0]).toEqual(bytes("3"));
      expect(results[1]).toBeNull();
      expect(results[2]).toEqual(bytes("1"));
      expect(results[3]).toEqual(bytes("3"));
    } finally {
      database.close();
    }
  });

  test("put many is atomic and delete many ignores missing", () => {
    const database = Database.create(temporaryDatabasePath());
    try {
      const entries: KvEntry[] = [
        { key: bytes("k1"), value: bytes("v1") },
        { key: bytes("k2"), value: bytes("v2") },
        { key: bytes("k3"), value: new Uint8Array(0) },
      ];
      database.kvPutMany(bytes("ns"), entries);
      expect(database.kvCount(bytes("ns"))).toBe(3n);

      database.kvDeleteMany(bytes("ns"), [bytes("k1"), bytes("ghost"), bytes("k3")]);
      expect(database.kvCount(bytes("ns"))).toBe(1n);
      expect(database.kvGet(bytes("ns"), bytes("k2"))).toEqual(bytes("v2"));

      database.kvPutMany(bytes("ns"), []);
      database.kvDeleteMany(bytes("ns"), []);
      expect(database.kvCount(bytes("ns"))).toBe(1n);
    } finally {
      database.close();
    }
  });

  test("partitions by namespace", () => {
    const database = Database.create(temporaryDatabasePath());
    try {
      database.kvPut(bytes("a"), bytes("key"), bytes("1"));
      database.kvPut(bytes("b"), bytes("key"), bytes("2"));

      expect(database.kvGet(bytes("a"), bytes("key"))).toEqual(bytes("1"));
      expect(database.kvGet(bytes("b"), bytes("key"))).toEqual(bytes("2"));
      expect(database.kvCount(bytes("a"))).toBe(1n);
      expect(database.kvCount(bytes("b"))).toBe(1n);

      database.kvClearNamespace(bytes("a"));
      expect(database.kvCount(bytes("a"))).toBe(0n);
      expect(database.kvCount(bytes("b"))).toBe(1n);
    } finally {
      database.close();
    }
  });
});

describe("AsyncDatabase key-value storage", () => {
  test("async crud and many", async () => {
    const database = AsyncDatabase.create(temporaryDatabasePath());

    await database.kvPut(bytes("cfg"), bytes("mode"), bytes("fast"));
    await database.kvPut(bytes("cfg"), bytes("limit"), bytes("10"));

    expect(await database.kvGet(bytes("cfg"), bytes("mode"))).toEqual(bytes("fast"));
    expect(await database.kvGet(bytes("cfg"), bytes("nope"))).toBeNull();
    expect(await database.kvCount(bytes("cfg"))).toBe(2n);

    const results = await database.kvGetMany(bytes("cfg"), [
      bytes("mode"),
      bytes("nope"),
      bytes("limit"),
    ]);
    expect(results.length).toBe(3);
    expect(results[0]).toEqual(bytes("fast"));
    expect(results[1]).toBeNull();
    expect(results[2]).toEqual(bytes("10"));

    const entries: KvEntry[] = [
      { key: bytes("x"), value: bytes("1") },
      { key: bytes("y"), value: bytes("2") },
    ];
    await database.kvPutMany(bytes("cfg"), entries);
    expect(await database.kvCount(bytes("cfg"))).toBe(4n);

    await database.kvDeleteMany(bytes("cfg"), [bytes("x"), bytes("nope"), bytes("y")]);
    expect(await database.kvCount(bytes("cfg"))).toBe(2n);

    await database.kvClearNamespace(bytes("cfg"));
    expect(await database.kvCount(bytes("cfg"))).toBe(0n);

    await database.close();
  });
});

File: bindings/go/kv_test.go
package zova

import (
	"bytes"
	"testing"
)

func TestKvCRUDPreservesExactBytes(t *testing.T) {
	db, err := Create(tempZovaPath(t, "kv"))
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	if err := db.KvPut([]byte("settings"), []byte("theme"), []byte("dark")); err != nil {
		t.Fatal(err)
	}
	if err := db.KvPut([]byte("settings"), []byte("retries"), []byte{0, 1, 2}); err != nil {
		t.Fatal(err)
	}
	if err := db.KvPut([]byte("settings"), []byte("empty"), nil); err != nil {
		t.Fatal(err)
	}

	value, err := db.KvGet([]byte("settings"), []byte("theme"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte("dark")) {
		t.Fatalf("KvGet theme = %q, want dark", value)
	}
	value, err = db.KvGet([]byte("settings"), []byte("retries"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte{0, 1, 2}) {
		t.Fatalf("KvGet retries = %v, want binary value", value)
	}
	value, err = db.KvGet([]byte("settings"), []byte("empty"))
	if err != nil {
		t.Fatal(err)
	}
	if value == nil || len(value) != 0 {
		t.Fatalf("KvGet empty = %v, want empty slice", value)
	}
	value, err = db.KvGet([]byte("settings"), []byte("ghost"))
	if err != nil {
		t.Fatal(err)
	}
	if value != nil {
		t.Fatalf("KvGet ghost = %q, want nil", value)
	}
	value, err = db.KvGet([]byte("other"), []byte("theme"))
	if err != nil {
		t.Fatal(err)
	}
	if value != nil {
		t.Fatalf("KvGet other/theme = %q, want nil", value)
	}

	count, err := db.KvCount([]byte("settings"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 3 {
		t.Fatalf("KvCount = %d, want 3", count)
	}

	if err := db.KvPut([]byte("settings"), []byte("theme"), []byte("light")); err != nil {
		t.Fatal(err)
	}
	value, err = db.KvGet([]byte("settings"), []byte("theme"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte("light")) {
		t.Fatalf("KvGet theme after put = %q, want light", value)
	}
	count, err = db.KvCount([]byte("settings"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 3 {
		t.Fatalf("KvCount after replace = %d, want 3", count)
	}

	if err := db.KvDelete([]byte("settings"), []byte("theme")); err != nil {
		t.Fatal(err)
	}
	value, err = db.KvGet([]byte("settings"), []byte("theme"))
	if err != nil {
		t.Fatal(err)
	}
	if value != nil {
		t.Fatalf("KvGet theme after delete = %q, want nil", value)
	}
	count, err = db.KvCount([]byte("settings"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 2 {
		t.Fatalf("KvCount after delete = %d, want 2", count)
	}

	if err := db.KvDelete([]byte("settings"), []byte("ghost")); err != nil {
		t.Fatal(err)
	}
	count, err = db.KvCount([]byte("settings"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 2 {
		t.Fatalf("KvCount after missing delete = %d, want 2", count)
	}

	if err := db.KvClearNamespace([]byte("settings")); err != nil {
		t.Fatal(err)
	}
	count, err = db.KvCount([]byte("settings"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 0 {
		t.Fatalf("KvCount after clear = %d, want 0", count)
	}
}

func TestKvGetManyPreservesOrderAndDuplicates(t *testing.T) {
	db, err := Create(tempZovaPath(t, "kv-many"))
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	for _, entry := range []struct{ key, value string }{
		{"a", "1"}, {"b", "2"}, {"c", "3"},
	} {
		if err := db.KvPut([]byte("ns"), []byte(entry.key), []byte(entry.value)); err != nil {
			t.Fatal(err)
		}
	}

	results, err := db.KvGetMany([]byte("ns"), [][]byte{[]byte("c"), []byte("ghost"), []byte("a"), []byte("c")})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 4 {
		t.Fatalf("KvGetMany len = %d, want 4", len(results))
	}
	assertKvResult(t, results[0], "3")
	if results[1] != nil {
		t.Fatalf("KvGetMany ghost = %v, want nil", *results[1])
	}
	assertKvResult(t, results[2], "1")
	assertKvResult(t, results[3], "3")
}

func TestKvPutManyAndDeleteMany(t *testing.T) {
	db, err := Create(tempZovaPath(t, "kv-batch"))
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	err = db.KvPutMany([]byte("ns"), []KvEntry{
		{Key: []byte("k1"), Value: []byte("v1")},
		{Key: []byte("k2"), Value: []byte("v2")},
		{Key: []byte("k3"), Value: nil},
	})
	if err != nil {
		t.Fatal(err)
	}
	count, err := db.KvCount([]byte("ns"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 3 {
		t.Fatalf("KvCount after put many = %d, want 3", count)
	}

	if err := db.KvDeleteMany([]byte("ns"), [][]byte{[]byte("k1"), []byte("ghost"), []byte("k3")}); err != nil {
		t.Fatal(err)
	}
	count, err = db.KvCount([]byte("ns"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 1 {
		t.Fatalf("KvCount after delete many = %d, want 1", count)
	}
	value, err := db.KvGet([]byte("ns"), []byte("k2"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte("v2")) {
		t.Fatalf("KvGet k2 = %q, want v2", value)
	}

	if err := db.KvPutMany([]byte("ns"), nil); err != nil {
		t.Fatal(err)
	}
	if err := db.KvDeleteMany([]byte("ns"), nil); err != nil {
		t.Fatal(err)
	}
	count, err = db.KvCount([]byte("ns"))
	if err != nil {
		t.Fatal(err)
	}
	if count != 1 {
		t.Fatalf("KvCount after empty batches = %d, want 1", count)
	}
}

func TestKvPartitionsByNamespace(t *testing.T) {
	db, err := Create(tempZovaPath(t, "kv-partition"))
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()

	if err := db.KvPut([]byte("a"), []byte("key"), []byte("1")); err != nil {
		t.Fatal(err)
	}
	if err := db.KvPut([]byte("b"), []byte("key"), []byte("2")); err != nil {
		t.Fatal(err)
	}

	value, err := db.KvGet([]byte("a"), []byte("key"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte("1")) {
		t.Fatalf("KvGet a/key = %q, want 1", value)
	}
	value, err = db.KvGet([]byte("b"), []byte("key"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(value, []byte("2")) {
		t.Fatalf("KvGet b/key = %q, want 2", value)
	}

	if err := db.KvClearNamespace([]byte("a")); err != nil {
		t.Fatal(err)
	}
	countA, err := db.KvCount([]byte("a"))
	if err != nil {
		t.Fatal(err)
	}
	countB, err := db.KvCount([]byte("b"))
	if err != nil {
		t.Fatal(err)
	}
	if countA != 0 || countB != 1 {
		t.Fatalf("after clear: a=%d b=%d, want 0 and 1", countA, countB)
	}
}

func assertKvResult(t *testing.T, result *[]byte, want string) {
	t.Helper()
	if result == nil {
		t.Fatalf("KvGetMany result is nil, want %q", want)
	}
	if !bytes.Equal(*result, []byte(want)) {
		t.Fatalf("KvGetMany result = %q, want %q", *result, want)
	}
}
